### PR TITLE
Make the player collisionbox settable

### PR DESCRIPTION
**Pull request by TeTpaAka**
Breaking changes with version 0.5

Reworked version of https://github.com/minetest/minetest/pull/2738.

Requires https://github.com/minetest/minetest_game/pull/1785 to work.

### CODE REVIEW

File: src/localplayer.h [hunk 1]
```diff
@@ -140,6 +140,8 @@ class LocalPlayer : public Player
 	v3f getEyePosition() const { return m_position + getEyeOffset(); }
 	v3f getEyeOffset() const;
 
+	void setCollisionbox(aabb3f box) { m_collisionbox = box; }
```
Comment (nerzhul) [inline src/localplayer.h]: const aabb3f &

Comment (TeTpaAka) [inline src/localplayer.h]: Fixed

**Comment by paramat** [comment]
Can this be merged soon or does it have to wait until freeze? Many use MTdev and are used to backwards compatibility. Just wondering.

**Comment by TeTpaAka** [comment]
This can't be merged soon. It completely breaks compatibility with old clients and old servers.

**Comment by nerzhul** [comment]
@TeTpaAka i think at a moment we should really inform user master will become unstable for server owners and they should stop to use it and go to stable-0.4

**Comment by paramat** [comment]
I agree with nerzhul, breaking compatibility earlier gives us more testing time before 0.5.
Once CSM restrictions are coded those can be backported to an 0.4 branch, togehter with any other useful bugfixes.

**Comment by nerzhul** [comment]
i will write a post on the forum soon and we can merge that PR if ready and approved at the end of the week, server owner should be careful

**Comment by paramat** [comment]
COncept was approved in #2738

**Comment by paramat** [comment]
We're ready to merge this as soon as it is rebased.

**Comment by TeTpaAka** [comment]
rebased

**Comment by paramat** [comment]
Thanks, will test.

**Comment by paramat** [comment]
screenshot_20170720_234436

screenshot_20170720_234119

screenshot_20170720_231918

Tested on a local server, Checked default collisionbox and custom collisionbox, checked default model has feet exactly at ground level.
:+1:

File: src/content_sao.cpp [hunk 2]
```diff
@@ -784,7 +784,7 @@ PlayerSAO::PlayerSAO(ServerEnvironment *env_, RemotePlayer *player_, u16 peer_id
 	m_prop.hp_max = PLAYER_MAX_HP;
 	m_prop.physical = false;
 	m_prop.weight = 75;
-	m_prop.collisionbox = aabb3f(-0.3f, -1.0f, -0.3f, 0.3f, 0.75f, 0.3f);
+	m_prop.collisionbox = aabb3f(-0.3f, 0.0f, -0.3f, 0.3f, 1.75f, 0.3f);
```
Comment (paramat) [inline src/content_sao.cpp]: On merge this can be changed to 1.77 to apply #6154

**Comment by paramat** [comment]
If someone other than me merges this please edit to 1.77f on merge to apply #6154

**Comment by paramat** [comment]
49920cfe8dc7869bb8235c6fe2b8ef26e2a523ca

**Comment by lhofhansl** [comment]
@TeTpaAka @SmallJoker @paramat this breaks various mods. At least:
- playeranim
- 3d armor
- MineClon2
- probably carbone, farlans, legend of minetest, etc. (they all override the character model)
- and probably many more

It's one thing to break client-server compatibility and also to break MTG only.
But breaking all these mods is perhaps not the best idea.

**Comment by octacian** [comment]
I honestly don't have a problem with it. Minetest is *not* supposed to be stable, if we never allow things like this to break then we'll never make any progress. For me anyway, it's kind of expected for things like this to break at times. In fact, my original expectation when I entered the modding community was that APIs would be broken all the time, not just like once every 2 years.

**Comment by nerzhul** [comment]
we are in 0.5 devel cycle development. Minetest is not stable and we are trying to fix very old issues which needs to break some things and mods can be broken and should be patched. Don't run 0.5 as a server except for debug/test purposes

**Comment by lhofhansl** [comment]
Fair enough.

**Comment by lhofhansl** [comment]
Can we give modders some tools to stay compatible if they wanted to?

Or perhaps instead of changing the models, the models would stay centered around the origin, and we apply a transformation upon loading...?

**Comment by paramat** [comment]
I disagree to both suggestions. Compatibility tools will delay modders updating their mods, which is what we want to happen as soon as possible.
A transformation is unnecessary complexity, 0.5 is a chance we have to break things in a way that also keeps them simple.
The point of breakage is that we don't add compatibility tools or keep backwards compatibility, if we did then it wouldn't be a breakage :]

**Comment by lhofhansl** [comment]
Sure. And I agree.

Why go out of the way to break things, though? What's better about model that is shifted up compared to one that is centered around the origin?

How does this work for other models (for mobs)? Are they centered around the origin?

Anyway, seems this might be better discussion on IRC.

**Comment by TeTpaAka** [comment]
All other entities have their origin at ground level. So this makes things more consistent.

The advantage is: Now you can use the same model for players and mobs. Also, the offset was weird to begin with. There were multiple locations, where the offset was applied, which makes the code hard to understand and difficult to maintain.

It wouldn't have been a problem to implement the collisionbox in a backward compatible way, but in the long run a consistent API is better than an API that never changes but is full of weird quirks.

**Comment by paramat** [comment]
^ yeah exactly that.

**Comment by lhofhansl** [comment]
Agreed. A standard model for models is worth breaking things. Thanks for bearing with me.

I'm just going to fix the mods that I use. Did playeranim and 3d-armor already. (see links here)

**Pull request closed without merging**